### PR TITLE
Prefer high bits

### PR DESCRIPTION
**Pull request by n8sh**
The output of some PRNGs has output that is higher quality in the the high bits than the low bits.

### CODE REVIEW

**Comment by n8sh** [comment]
Unfortunately this wasn't entirely disentangled from https://github.com/libmir/mir-random/pull/35 since the generator introduced there is one of the generators I wanted to annotate.

**Comment by n8sh** [comment]
Restricted unittest to version(mir_random_test).

File: source/mir/random/package.d [hunk 1]
```diff
@@ -89,7 +116,8 @@ Returns:
 bool rand(T : bool, G)(ref G gen)
     if (isSaturatedRandomEngine!G)
 {
-    return gen() & 1;
+    import std.traits : Unsigned;
+    return 0 > cast(Unsigned!(EngineReturnType!G)) gen();
```
Comment (9il) [inline source/mir/random/package.d]: `Signed!` ?

Comment (n8sh) [reply source/mir/random/package.d]: OOPS! Yes, signed! Very important mistake.

**Comment by 9il** [comment]
Needs rebase

File: source/mir/random/package.d [hunk 2]
```diff
@@ -47,6 +47,28 @@ else
...
 /++
+Are the high bits of the engine's output known to have
+better statistical properties than the low bits of the
+output?
+
+True for:
+* linear congruential generators with power-of-2 modulus
+* xorshift+ family
+* xorshift* family
+* in principle any generator whose final operation is something like
```
Comment (9il) [inline source/mir/random/package.d]: Does `* ` works for DDOC?

File: source/mir/random/package.d [hunk 3]
```diff
@@ -47,6 +47,28 @@ else
...
+multiplication or addition in which the high bits depend on the low bits
+but the low bits are unaffected by the high bits.
++/
+//If we make this public instead of private it should probably go in mir.random.engine.package.
```
Comment (9il) [inline source/mir/random/package.d]: Yes, please move it and make public

File: source/mir/random/package.d [hunk 4]
```diff
@@ -47,6 +47,28 @@ else
...
+private template preferHighBits(G)
+{
```
Comment (9il) [inline source/mir/random/package.d]: if (isSaturatedRandomEngine!G)

**Pull request merged**